File: src/behavioral/command.java
public class Room {
    private Light light;

    public Room() {
        this.light = new Light();
    }

    public void switchLights() {
        light.setSwitchedOn(!light.isSwitchedOn());
    }
}

public class Kitchen extends Room {
    private Oven oven;
    //
}

public class Bathroom extends Room {
    private String hotWater;
    //
}

public class LivingRoom extends Room {
    private Windows windows;
    //
}

public class Bedroom extends Room {
    private String music;
    //
}

public class Light {
    private boolean switchedOn;

    public boolean isSwitchedOn() {
        return switchedOn;
    }

    public void setSwitchedOn(boolean switchedOn) {
        this.switchedOn = switchedOn;
    }
}

public class House {
    List<Room> rooms;

    public House() {
        rooms = new ArrayList<>();
    }

    public void addRoom(Room room) {
        rooms.add(room);
    }
}

public static void main(String[] args) {
    House house = new House();
    house.addRoom(new LivingRoom());
    house.addRoom(new Bathroom());
    house.addRoom(new Kitchen());
    house.addRoom(new Bedroom());
    house.addRoom(new Bedroom());
    house.rooms.forEach(Room::switchLights);
}

File: src/behavioral/command.ts
class Room {
  private light: Light;

  constructor() {
    this.light = new Light();
  }

  switchLights(): void {
    this.light.setSwitchedOn(!this.light.isSwitchedOn());
  }
}

interface Oven {}
interface Windows {}

class Kitchen extends Room {
  public readonly oven!: Oven;

  //
}

class Bathroom extends Room {
  public readonly hotWater!: string;

  //
}

class LivingRoom extends Room {
  public readonly windows!: Windows;

  //
}

class Bedroom extends Room {
  public readonly music!: string;

  //
}

class Light {
  private switchedOn: boolean = false;

  isSwitchedOn(): boolean {
    return this.switchedOn;
  }

  setSwitchedOn(switchedOn: boolean): void {
    console.log({ switchedOn });
    this.switchedOn = switchedOn;
  }
}

class House {
  rooms: Room[] = [];

  addRoom(room: Room): void {
    this.rooms.push(room);
  }
}

function smartHouse(): void {
  const house = new House();
  house.addRoom(new LivingRoom());
  house.addRoom(new Bathroom());
  house.addRoom(new Kitchen());
  house.addRoom(new Bedroom());
  house.addRoom(new Bedroom());

  console.log({ house });

  house.rooms.forEach((room) => {
    room.switchLights();
  });
}

smartHouse();
